File: app/src/main/java/com/adsamcik/signalcollector/fragments/FragmentActivities.java
package com.adsamcik.signalcollector.fragments;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.adsamcik.signalcollector.R;
import com.adsamcik.signalcollector.data.Challenge;
import com.adsamcik.signalcollector.interfaces.ITabFragment;
import com.adsamcik.signalcollector.utility.Assist;
import com.adsamcik.signalcollector.utility.ChallengeManager;
import com.adsamcik.signalcollector.utility.Failure;
import com.adsamcik.signalcollector.utility.SnackMaker;

public class FragmentActivities extends Fragment implements ITabFragment {
	private ListView listViewChallenges;
	private SwipeRefreshLayout refreshLayout;

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View rootView = inflater.inflate(R.layout.fragment_activities, container, false);
		final Activity activity = getActivity();

		listViewChallenges = rootView.findViewById(R.id.listview_challenges);

		refreshLayout = (SwipeRefreshLayout) rootView;
		refreshLayout.setColorSchemeResources(R.color.color_primary);
		refreshLayout.setProgressViewOffset(true, 0, Assist.dpToPx(activity, 40));
		refreshLayout.setOnRefreshListener(this::updateData);

		updateData();

		return rootView;
	}

	private void updateData() {
		final boolean isRefresh = refreshLayout != null && refreshLayout.isRefreshing();
		final Activity activity = getActivity();
		final Context context = activity.getApplicationContext();
		ChallengeManager.getChallenges(activity, isRefresh, (source, challenges) -> {
			if (!source.isSuccess())
				new SnackMaker(activity).showSnackbar(R.string.error_connection_failed);
			else {
				activity.runOnUiThread(() -> listViewChallenges.setAdapter(new ChallengesAdapter(context, challenges)));
			}
			activity.runOnUiThread(() -> refreshLayout.setRefreshing(false));
		});
	}

	@NonNull
	@Override
	public Failure<String> onEnter(@NonNull FragmentActivity activity, @NonNull FloatingActionButton fabOne, @NonNull FloatingActionButton fabTwo) {
		return new Failure<>();
	}

	@Override
	public void onLeave(@NonNull FragmentActivity activity) {

	}

	@Override
	public void onPermissionResponse(int requestCode, boolean success) {

	}

	@Override
	public void onHomeAction() {

	}

	private class ChallengesAdapter extends BaseAdapter {
		private final Context mContext;
		private final LayoutInflater mInflater;
		private final Challenge[] mDataSource;

		public ChallengesAdapter(Context context, Challenge[] items) {
			mContext = context;
			mDataSource = items;
			mInflater = (LayoutInflater) mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		}

		@Override
		public int getCount() {
			return mDataSource.length;
		}

		@Override
		public Object getItem(int i) {
			return mDataSource[i];
		}

		@Override
		public long getItemId(int i) {
			return i;
		}

		@Override
		public View getView(int i, View view, ViewGroup viewGroup) {
			if (view == null)
				view = mInflater.inflate(R.layout.layout_challenge_small, viewGroup, false);

			Challenge challenge = mDataSource[i];
			((TextView) view.findViewById(R.id.challenge_title)).setText(challenge.getTitle());
			((TextView) view.findViewById(R.id.challenge_description)).setText(challenge.getDescription());

			TextView textViewDifficulty = view.findViewById(R.id.challenge_difficulty);
			if(challenge.getDifficultyString() == null)
				textViewDifficulty.setVisibility(View.GONE);
			else
				textViewDifficulty.setText(challenge.getDifficultyString());


			Context context = getContext();

			int color;
			if(challenge.isDone()) {
				color = ContextCompat.getColor(getContext(), R.color.background_success);
			} else {
				TypedValue typedValue = new TypedValue();
				context.getTheme().resolveAttribute(android.R.attr.colorBackground, typedValue, true);
				color = typedValue.data;
			}
			view.setBackgroundColor(color);
			return view;
		}
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/fragments/FragmentActivities.kt
package com.adsamcik.signalcollector.fragments

import androidx.appcompat.app.AppCompatActivity
import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListView
import android.widget.TextView
import com.adsamcik.draggable.IOnDemandView
import com.adsamcik.signalcollector.R
import com.adsamcik.signalcollector.data.Challenge
import com.adsamcik.signalcollector.enums.ChallengeDifficulties
import com.adsamcik.signalcollector.extensions.dpAsPx
import com.adsamcik.signalcollector.interfaces.IViewChange
import com.adsamcik.signalcollector.test.useMock
import com.adsamcik.signalcollector.uitools.ColorManager
import com.adsamcik.signalcollector.uitools.ColorSupervisor
import com.adsamcik.signalcollector.uitools.ColorView
import com.adsamcik.signalcollector.utility.ChallengeManager
import com.adsamcik.signalcollector.utility.SnackMaker
import kotlinx.android.synthetic.main.fragment_activities.*
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.launch
import kotlin.math.roundToInt

class FragmentActivities : androidx.fragment.app.Fragment(), IOnDemandView {
    private lateinit var listViewChallenges: ListView
    private lateinit var refreshLayout: androidx.swiperefreshlayout.widget.SwipeRefreshLayout

    private lateinit var colorManager: ColorManager

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_activities, container, false)

        listViewChallenges = rootView.findViewById(R.id.listview_challenges)

        refreshLayout = rootView.findViewById(R.id.swiperefresh_activites)
        refreshLayout.setColorSchemeResources(R.color.color_primary)
        refreshLayout.setProgressViewOffset(true, 0, 40.dpAsPx)
        refreshLayout.setOnRefreshListener({ this.updateData() })

        updateData()

        val context = context!!
        listViewChallenges.adapter = ChallengesAdapter(context, arrayOf())
        colorManager = ColorSupervisor.createColorManager(context)
        colorManager.watchAdapterView(ColorView(listViewChallenges, 1, true, false))

        return rootView
    }

    private fun updateData() {
        val isRefresh = refreshLayout.isRefreshing
        val activity = activity!!
        val context = activity.applicationContext
        if (useMock) {
            val challenges = arrayOf(Challenge(Challenge.ChallengeType.AIsForAlphabet, "Hello world!", arrayOf("20", "h"), 0.5f, ChallengeDifficulties.MEDIUM),
                    Challenge(Challenge.ChallengeType.AwfulExplorer, "Hello world!", arrayOf("6000"), 0f, ChallengeDifficulties.HARD),
                    Challenge(Challenge.ChallengeType.Crowded, "Hello world!", arrayOf("50"), 1f, ChallengeDifficulties.EASY))

            challenges.forEach { it.generateTexts(context) }
            launch(UI) {
                (listViewChallenges.adapter as ChallengesAdapter).updateData(challenges)
                refreshLayout.isRefreshing = false
            }
        } else {
            launch {
                val (source, challenges) = ChallengeManager.getChallenges(activity, isRefresh)
                if (!source.success)
                    SnackMaker(activities_root).showSnackbar(R.string.error_connection_failed)
                else {
                    launch(UI) {
                        (listViewChallenges.adapter as ChallengesAdapter).updateData(challenges!!)
                    }
                }
                launch(UI) { refreshLayout.isRefreshing = false }
            }
        }
    }

    override fun onEnter(activity: AppCompatActivity) {}

    override fun onLeave(activity: AppCompatActivity) {}

    override fun onPermissionResponse(requestCode: Int, success: Boolean) {

    }

    private inner class ChallengesAdapter(mContext: Context, private var mDataSource: Array<Challenge>) : BaseAdapter(), IViewChange {
        private val mInflater: LayoutInflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        fun updateData(challenges: Array<Challenge>) {
            this.mDataSource = challenges
            notifyDataSetInvalidated()
        }

        override var onViewChangedListener: ((View) -> Unit)? = null

        override fun getCount(): Int = mDataSource.size

        override fun getItem(i: Int): Any = mDataSource[i]

        override fun getItemId(i: Int): Long = i.toLong()

        override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
            var fragmentView = view
            if (fragmentView == null)
                fragmentView = mInflater.inflate(R.layout.layout_challenge_small, viewGroup, false)

            val challenge = mDataSource[i]
            (fragmentView!!.findViewById<View>(R.id.challenge_title) as TextView).text = challenge.title
            (fragmentView.findViewById<View>(R.id.challenge_description) as TextView).text = challenge.description

            val textViewDifficulty = fragmentView.findViewById<TextView>(R.id.challenge_difficulty)
            if (challenge.difficultyString == null)
                textViewDifficulty.visibility = View.GONE
            else
                textViewDifficulty.text = challenge.difficultyString

            (fragmentView.findViewById<View>(R.id.challenge_progress) as TextView).text = getString(R.string.challenge_progress, (challenge.progress * 100).toInt())
            val color = ColorUtils.setAlphaComponent(ContextCompat.getColor(context!!, R.color.background_success), (challenge.progress * 254).roundToInt())
            fragmentView.setBackgroundColor(color)

            onViewChangedListener?.invoke(fragmentView)
            return fragmentView
        }
    }
}
